ember-cli/2017-03/march-31.md

```diff
@@ -0,0 +1,127 @@
+  - `this.import` is relative to whatever `this` is
+- path *has* to start with `vendor`, `node_modules`
+
+### Blrpnt extraction
+
+- unfork
+  - port new test changes from ember-cli
+  - make sure we haven't lost anything added in ember-cli/blueprint.js
+- load blprnt in ember-cli
+  - bring into ember-cli as drop in replacement
+- decouple pods/classic structures information from blprint/blueprint.js
+  - create new structures using new primitive to manage this logic in ember-cli
+- create module unification structure to provide blueprint support
+  - use glimmer resolver to take advantage of rules to generate file paths
+
+### Background Service
+
+**idea 1**: Convert bin/cli into an actor system to provide parallelism  (actor === task)
+
+**idea 2**: Daemonize the process
+
+- removing add-ons (packages.json changes) needs to restart daemon
+- daemon per application
+- abstraction at app layer
+
+*Conclusion*: Since the gain of this refactor doesn't seem high enough
+at this point we will not move forward with those plans.
+
+### Slow Commands
+
+#### `generate`:
+
+- idea 1: cache the project property that gets passed in, which is probably the biggest slowdown
+- idea 2: based on ember-cli version, add-on commands will use a shim for `includedCommands`
+  to export a hash with keys being the command name, and value being a hash including the path
+  to the module to be required, as well as command options and any other relevant info
+  - the primary issue comes from requires at the top of the command
+- deprecate command name property, in favor of the key
+
+#### `help`
+
+- upon add-on install, identify and write an add-on's available blueprint g -h json
+  output to a per project file like `.ember-cli-help`  or something similar
+- `ember -h` would parse that generated json file and generate the terminal output
+
+### Ember CLI/Broccoli Errors
+
+Currently, our error handling in ember cli is very rudimentary (i.e. Template Compilation errors
+when all that uses see is confusing message from Glimmer internals). There were previous [efforts](https://github.com/ember-cli/broccoli-middleware/pull/4) to outline what it might look like ideally.
+
+This would require:
+
+- changes to handlebars compiler itself (it has error information already, just doesn't propagate it up);
+- changes to ember-template-compiler
+- changes to glimmer
+- changes to broccoli-middleware
+- changes to broccoli-builder (this one is interesting b/c we diverged from the mainstream version;
+  mainstream version does some work to expose error information already; elephant in the room is should
+  be un-fork and start using mainstream one or just keep using the fork)
+- creating a package that would unify error information across projects mentioned above,
+  a.k.a `new BroccoliError(/*some args*/)`
+- some design work to make it nice
+- profit?
+
+### Global Add-ons
+
+- Needs an RFC
+- Save global add-ons in user home folder (like `.ember-cli`)
```

Comment:
Although possible, this will require care. As the compatibility story is a tad complicated, for example old and new versions of ember-cli should not fail because some global addon is not compatible with them.

Likely just some capability handshake (or version check) is required.